src/status_im/contexts/centralized_metrics/tracking.cljs

```diff
@@ -43,15 +43,31 @@
     :screen/onboarding.syncing-progress
     :screen/onboarding.syncing-progress-intro
     :screen/onboarding.syncing-results
-    :screen/onboarding.welcome})
+    :screen/onboarding.welcome
+
+    ;; Collectibles
+    :screen/wallet.collectible})
 
 (defn track-view-id-event
   [view-id]
   (when (contains? view-ids-to-track view-id)
     (navigation-event (name view-id))))
 
+(defn collectilbes-fetched-event
+  [db]
+  (let [accounts               (get-in db [:wallet :accounts])
+        amount-on-all-accounts (reduce (fn [collectibles-amount account]
+                                         (+ collectibles-amount (:current-collectible-idx account)))
+                                       0
+                                       (vals accounts))]
+    (key-value-event "collectibles-fetched" :total-amount amount-on-all-accounts)))
+
+(defn navigated-to-collectibles-tab-event
+  [location]
+  (key-value-event "navigated-to-collectibles-tab" :location location))
+
 (defn tracked-event
-  [[event-name second-parameter]]
+  [[event-name second-parameter] db]
```

Comment:
I'm not sure this is the right way of extending this API, it is becoming more restrictive instead of more flexible.

IMO, we could create a two-arity version instead, since with this change, we call it only once as:

`(tracked-event '[...] db)`

And multiple times as:

`(tracked-event '[...] {})`


And... actually, this `second-parameter` value is also not passed to the function call sometimes 🤔
I think the signature could have been the following since the beginning:

```clojure
(defn tracked-event
  [event-name & {:keys[...] :as opts}]
  ...)
```

CC: @ilmotta

Reply:
Thanks for tagging me @ulisesmac. Below I share a few roughly sketched ideas, not blocking the PR from my PoV. Text is long, hard to explain, but I hope it makes some sense.

I think it's not the best that the tracking code now has access to the entire app-db and knows the inner details of its structure. Better I would say we should give the least amount of data to this layer and keep it plain simple, also because it's not well QAed.

If we need to compute something from the app-db that needs to be tracked, and which is decoupled from the metrics layer we could:

1. Define a root key in the app-db, for example, `:centralized-metrics/event-data`, which will store any arbitrary data that's safe to be tracked and which is serializable to JSON.
2. The event that wants to be tracked can `assoc` to this root key. In this case, in event `:wallet/flush-collectibles-fetched` it would `assoc` the total number of collectibles.
3. The interceptor `centralized-metrics-interceptor` will then extract the value of `:centralized-metrics/event-data` and pass it to `tracking/tracked-event`.

With this solution, the centralized metrics layer will not know _how_ to compute anything, but will still be able to get computed track data and/or arguments passed to any re-frame event. This would scale to any other event in re-frame and we would still keep this layer simple/dumb.

-----

About the signature of the function `tracked-event`, an alternative is to implement `tracked-event` with the `case` macro, but the `event` argument (first arg to `tracked-event`) would be passed to downstream functions, and then we would only destructure the re-frame event in each specific function signature, thus solving the naming problem of `second-parameter` or how to best destructure the event because each function will destructure and name the arguments knowing what they really are.

In this example, there's no generic name like `second-parameter` and we don't need to worry about the full signature of the re-frame event:

```clojure
(defn- metrics-enabled-event
  [[_ enabled?]]
  (key-value-event "events.metrics-enabled" :enabled enabled?))

(defn tracked-event
  [[event-name :as event] event-data]
  (case event-name
    :profile/get-profiles-overview-success
    (user-journey-event app-started-event)

    :centralized-metrics/toggle-centralized-metrics
    (metrics-enabled-event event)

    :set-view-id
    (track-view-id-event event)

    :wallet/select-account-tab
    (navigated-to-collectibles-tab-event event)

    :wallet/select-home-tab
    (navigated-to-collectibles-tab-event event)

    ;; => NOTE: Here `event-data` was computed outside the centralized metrics layer and was extracted by the global interceptor when calling tracked-event 
    :wallet/flush-collectibles-fetched
    (collectilbes-fetched-event event-data)

    nil))
```

Reply:
> Thanks for tagging me @ulisesmac. Below I share a few roughly sketched ideas, not blocking the PR from my PoV. Text is long, hard to explain, but I hope it makes some sense.
> 
> I think it's not the best that the tracking code now has access to the entire app-db and knows the inner details of its structure. Better I would say we should give the least amount of data to this layer and keep it plain simple, also because it's not well QAed.
> 
> If we need to compute something from the app-db that needs to be tracked, and which is decoupled from the metrics layer we could:
> 
> 1. Define a root key in the app-db, for example, `:centralized-metrics/event-data`, which will store any arbitrary data that's safe to be tracked and which is serializable to JSON.
> 2. The event that wants to be tracked can `assoc` to this root key. In this case, in event `:wallet/flush-collectibles-fetched` it would `assoc` the total number of collectibles.
> 3. The interceptor `centralized-metrics-interceptor` will then extract the value of `:centralized-metrics/event-data` and pass it to `tracking/tracked-event`.
> 
> With this solution, the centralized metrics layer will not know _how_ to compute anything, but will still be able to get computed track data and/or arguments passed to any re-frame event. This would scale to any other event in re-frame and we would still keep this layer simple/dumb.
> 
> -----
> 
> About the signature of the function `tracked-event`, an alternative is to implement `tracked-event` with the `case` macro, but the `event` argument (first arg to `tracked-event`) would be passed to downstream functions, and then we would only destructure the re-frame event in each specific function signature, thus solving the naming problem of `second-parameter` or how to best destructure the event because each function will destructure and name the arguments knowing what they really are.
> 
> In this example, there's no generic name like `second-parameter` and we don't need to worry about the full signature of the re-frame event:
> 
> ```clojure
> (defn- metrics-enabled-event
>   [[_ enabled?]]
>   (key-value-event "events.metrics-enabled" :enabled enabled?))
> 
> (defn tracked-event
>   [[event-name :as event] event-data]
>   (case event-name
>     :profile/get-profiles-overview-success
>     (user-journey-event app-started-event)
> 
>     :centralized-metrics/toggle-centralized-metrics
>     (metrics-enabled-event event)
> 
>     :set-view-id
>     (track-view-id-event event)
> 
>     (:wallet/select-account-tab :wallet/select-home-tab)
>     (navigated-to-collectibles-tab-event event)
> 
>     ;; => NOTE: Here `event-data` was computed outside the centralized metrics layer and was extracted by the global interceptor when calling tracked-event 
>     :wallet/flush-collectibles-fetched
>     (collectilbes-fetched-event event-data)
> 
>     nil))
> ```

Thanks for sharing your thoughts @ilmotta

I agree with you about not passing `db` to this event. I'm still not sure about growing app-db horizontally instead of nested, but that's a conversation for another moment.

Reply:
@ilmotta, I have one more thought. Personally I would be happy to avoid both - referencing `db` from metrics namespace and having a separate key in `re-frame` db, because it should be not only set but also cleaned after every event to make sure that intercepted event didn't get wrong data from previous event. And some events just don't need it.

The way that seems more clean for me - is having dedicated event every time when we need some additional data to be computed from `db` and intercepting that event.
For example in current case when `:wallet/flush-collectibles-fetched` event fired it can dispatch one more event - `:metrics/track-user-has-collectibles?` which receives parameter `:has-collectibles?` as an argument and this event we can intercept from metrics layer.

So `flush-collectibles` will become:

```clojure
(defn flush-collectibles
  [{:keys [db]}]
  (let [collectibles-per-account (get-in db [:wallet :ui :collectibles :fetched])
        accounts               (get-in db [:wallet :accounts])
        has-collectibles? (some #(pos? (:current-collectible-idx %)) (vals accounts))]
    {:db (-> db
             (update-in [:wallet :ui :collectibles] dissoc :pending-requests :fetched)
             (update-in [:wallet :accounts] move-collectibles-to-accounts collectibles-per-account))
     :fx [:dispatch
          [:metrics/track-user-has-collectibles? has-collectibles?]]}))
```

We discussed with @shivekkhurana that he would prefer not to pollute our events code with metrics-related calculations but if we anyway need some calculations to be done, it better be done not in the metrics namespace as you suggested, and i think it is better to be in a form of separate event.
Wdyt?

Reply:
> because it should be not only set but also cleaned after every event

Good point about clean-up. We can use the interceptor itself to clean up because the `context` we receive has the `:db` effect and we can grab the value of `:centralized-metrics/event-data`, use it to generate the MixPanel event and then return a new context with the `:db` effect dissoc'ing `:centralized-metrics/event-data` data key. The result is that nothing will leak to another event and, actually, the app-db won't be modified to persist the computed event data. No subscription recomputations should happen if we do this correctly.

-----

I also agree with @shivekkhurana that we shouldn't pollute normal code with metrics code and that it's highly desirable to keep the metrics layer dumb.

I think the idea of using separate events can work well 👍🏼, but I would personally use that only when necessary because of the overhead and extra verbosity. For example, in the example you provided, a simple `assoc` gets the job done and it's decent for readability. Also considering the clean-up will be automatic by the interceptor. 

Including event data computation in the tracked re-frame event is not bad actually, there's an advantage, the computation can be unit tested with the rest of the handler, which is actually quite important because if we start to track incorrect computations it will really complicate analysis in MixPanel and I doubt QAs will have the time to manually double-check metrics are generated correctly. Actually unfair to ask them to do that.

```clojure
(defn flush-collectibles
  [{:keys [db]}]
  (let [collectibles-per-account (get-in db [:wallet :ui :collectibles :fetched])
        accounts                 (get-in db [:wallet :accounts])
        has-collectibles?        (some #(pos? (:current-collectible-idx %)) (vals accounts))]
    {:db (-> db
             (update-in [:wallet :ui :collectibles] dissoc :pending-requests :fetched)
             (update-in [:wallet :accounts] move-collectibles-to-accounts collectibles-per-account)
             (assoc :centralized-metrics/event-data has-collectibles?))}))
```